### ISSUE TITLE
Assertion failure on trivial program.

### ISSUE THREAD

**Issue by zentner-kyle** [mozilla/rr#1531]
I get the following error message:
```
[FATAL /home/zentner/rr/src/ReplaySession.cc:915:check_ticks_consistency() errno: 0 'Success']
 (task 29659 (rec:28718) at time 170)
 ->  Assertion `llabs(ticks_now - trace_ticks) <= ticks_slack' failed to hold. ticks mismatch for 'PATCH_SYSCALL'; expected 45407, got 46688
```

Every time I attempt to replay the following program (a copy of test32.c):
```
int main() { return 0; }
```
Which I've compiled using:
```
gcc test.c -o test
```

I'm on the most up to date commit (868ff270971cde560e73bbb9a77b184b3ffc2802), but this seems to occur on all commit I try.

I'm on ArchLinux, and can reproduce the issue on both the main (4.1.2) and lts (3.14) kernels.

`uname -a`:
```
Linux lili 4.1.2-2-ARCH #1 SMP PREEMPT Wed Jul 15 08:30:32 UTC 2015 x86_64 GNU/Linux
```

Hardware summary (from `lshw`):
```
    description: Notebook
    product: MacBookPro11,3 (System SKU#)
    vendor: Apple Inc.
    version: 1.0
    serial: C02PG3STG3QP
    width: 64 bits
    capabilities: smbios-2.4 dmi-2.4 vsyscall32
    configuration: boot=normal chassis=notebook family=Mac sku=System SKU# uuid=59848856-1A44-335C-B5A8-A459102E7789
```
`lshw -c cpu`:
```
  *-cpu
       description: CPU
       product: Core i7
       vendor: Intel Corp.
       physical id: 0
       bus info: cpu@0
       version: Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
       serial: To Be Filled By O.E.M.
       slot: U3E1
       size: 802MHz
       capacity: 4GHz
       width: 64 bits
       clock: 25MHz
       capabilities: fpu fpu_exception wp vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp x86-64 constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm ida arat epb pln pts dtherm tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm xsaveopt cpufreq
```

`dmesg | grep PMU`:
```
[    0.069315] Performance Events: PEBS fmt2+, 16-deep LBR, Haswell events, full-width counters, Intel PMU driver.
[    0.106419] NMI watchdog: enabled on all CPUs, permanently consumes one hw-PMU counter.
[    0.326660] perf_event_intel: PMU erratum BJ122, BV98, HSD29 worked around, HT is on
```

**Comment by rocallahan** [comment]
Ouch!

What's in `rr dump` around global_time:170?

**Comment by zentner-kyle** [comment]
The following:
```
{
  global_time:160, event:`SYSCALL: fcntl' (state:ENTERING_SYSCALL) tid:28090, ticks:44155
 rax:0xffffffffffffffda rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x7fffffffe1f0 rsi:0xf rdi:0x64 rbp:0x64 rsp:0x7fffffffe140 r8:0x0 r9:0x0 r10:0x0 r11:0x246 r12:0x7fffffffe200 r13:0x3 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x48
}
{
  global_time:161, event:`SYSCALL: fcntl' (state:EXITING_SYSCALL) tid:28090, ticks:44155
 rax:0x0 rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x7fffffffe1f0 rsi:0xf rdi:0x64 rbp:0x64 rsp:0x7fffffffe140 r8:0x0 r9:0x0 r10:0x0 r11:0x246 r12:0x7fffffffe200 r13:0x3 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x48
}
{
  global_time:162, event:`SYSCALL: fcntl' (state:ENTERING_SYSCALL) tid:28090, ticks:44157
 rax:0xffffffffffffffda rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x1e rsi:0xa rdi:0x64 rbp:0x64 rsp:0x7fffffffe140 r8:0x0 r9:0x0 r10:0x0 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x48
}
{
  global_time:163, event:`SYSCALL: fcntl' (state:EXITING_SYSCALL) tid:28090, ticks:44157
 rax:0x0 rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x1e rsi:0xa rdi:0x64 rbp:0x64 rsp:0x7fffffffe140 r8:0x0 r9:0x0 r10:0x0 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x48
}
{
  global_time:164, event:`SYSCALL: rt_sigprocmask' (state:ENTERING_SYSCALL) tid:28090, ticks:44160
 rax:0xffffffffffffffda rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x7fffffffe200 rsi:0x7fffffffe150 rdi:0x0 rbp:0x7fffffffe200 rsp:0x7fffffffe100 r8:0x0 r9:0x0 r10:0x8 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe
}
{
  global_time:165, event:`SYSCALL: rt_sigprocmask' (state:EXITING_SYSCALL) tid:28090, ticks:44160
 rax:0x0 rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x7fffffffe200 rsi:0x7fffffffe150 rdi:0x0 rbp:0x7fffffffe200 rsp:0x7fffffffe100 r8:0x0 r9:0x0 r10:0x8 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe
}
{
  global_time:166, event:`SYSCALL: rrcall_init_buffers' (state:ENTERING_SYSCALL) tid:28090, ticks:44161
 rax:0xffffffffffffffda rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x0 rsi:0x0 rdi:0x7fffffffe1f0 rbp:0x64 rsp:0x7fffffffe1a0 r8:0x0 r9:0x0 r10:0x0 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x1bb
}
{
  global_time:167, event:`SYSCALL: rrcall_init_buffers' (state:EXITING_SYSCALL) tid:28090, ticks:44161
 rax:0x2aaaaab08000 rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x0 rsi:0x0 rdi:0x7fffffffe1f0 rbp:0x64 rsp:0x7fffffffe1a0 r8:0x0 r9:0x0 r10:0x0 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x1bb
}
{
  global_time:168, event:`SYSCALL: rt_sigprocmask' (state:ENTERING_SYSCALL) tid:28090, ticks:44162
 rax:0xffffffffffffffda rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x0 rsi:0x7fffffffe200 rdi:0x2 rbp:0x64 rsp:0x7fffffffe1a0 r8:0x0 r9:0x0 r10:0x8 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe
}
{
  global_time:169, event:`SYSCALL: rt_sigprocmask' (state:EXITING_SYSCALL) tid:28090, ticks:44162
 rax:0x0 rbx:0xfffffffffffffff8 rcx:0xffffffffffffffff rdx:0x0 rsi:0x7fffffffe200 rdi:0x2 rbp:0x64 rsp:0x7fffffffe1a0 r8:0x0 r9:0x0 r10:0x8 r11:0x246 r12:0x7fffffffe200 r13:0x7fffffffe1f0 r14:0x6dba r15:0x0 rip:0x70000038 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe
}
{
  global_time:170, event:`PATCH_SYSCALL' tid:28090, ticks:46688
}
{
  global_time:171, event:`SYSCALLBUF_FLUSH' tid:28090, ticks:46716
}
{
  global_time:172, event:`PATCH_SYSCALL' tid:28090, ticks:46716
}
{
  global_time:173, event:`SYSCALLBUF_RESET' tid:28090, ticks:46716
}
{
  global_time:174, event:`SYSCALLBUF_FLUSH' tid:28090, ticks:46780
}
{
  global_time:175, event:`PATCH_SYSCALL' tid:28090, ticks:46780
}
{
  global_time:176, event:`SYSCALLBUF_RESET' tid:28090, ticks:46780
}
{
  global_time:177, event:`SYSCALL: mmap' (state:ENTERING_SYSCALL) tid:28090, ticks:46787
 rax:0xffffffffffffffda rbx:0x9 rcx:0xffffffffffffffff rdx:0x1 rsi:0x188820 rdi:0x0 rbp:0x7fffffffe170 rsp:0x7fffffffddd0 r8:0x3 r9:0x0 r10:0x2 r11:0x202 r12:0x188820 r13:0x7fffffffe2b0 r14:0x7fffffffe020 r15:0x7fffffffded0 rip:0x70000018 eflags:0x202 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x9
}
{
  global_time:178, event:`SYSCALL: mmap' (state:EXITING_SYSCALL) tid:28090, ticks:46787
 rax:0x2aaaab6ac000 rbx:0x9 rcx:0xffffffffffffffff rdx:0x1 rsi:0x188820 rdi:0x0 rbp:0x7fffffffe170 rsp:0x7fffffffddd0 r8:0x3 r9:0x0 r10:0x2 r11:0x202 r12:0x188820 r13:0x7fffffffe2b0 r14:0x7fffffffe020 r15:0x7fffffffded0 rip:0x70000018 eflags:0x202 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x9
}
{
  global_time:179, event:`SYSCALL: close' (state:ENTERING_SYSCALL) tid:28090, ticks:46792
 rax:0xffffffffffffffda rbx:0x7fffffffe020 rcx:0xffffffffffffffff rdx:0x2ac7 rsi:0x188820 rdi:0x3 rbp:0x7fffffffe170 rsp:0x7fffffffe020 r8:0x3 r9:0x0 r10:0x2 r11:0x216 r12:0x188820 r13:0x7fffffffe2b0 r14:0x2aaaab6ac000 r15:0x3 rip:0x2aaaaaf1448d eflags:0x216 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x3
}
{
  global_time:180, event:`SYSCALL: close' (state:EXITING_SYSCALL) tid:28090, ticks:46792
 rax:0x0 rbx:0x7fffffffe020 rcx:0xffffffffffffffff rdx:0x2ac7 rsi:0x188820 rdi:0x3 rbp:0x7fffffffe170 rsp:0x7fffffffe020 r8:0x3 r9:0x0 r10:0x2 r11:0x216 r12:0x188820 r13:0x7fffffffe2b0 r14:0x2aaaab6ac000 r15:0x3 rip:0x2aaaaaf1448d eflags:0x216 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x3
}
```

**Comment by rocallahan** [comment]
If you replay up to event 169 (`rr replay -g 169`) and then in gdb print the values of `__pthread_force_elision`, `__elision_available` and `__elision_aconf`, what do you get? For the latter you might need to manually dump all 16 bytes of it.

**Comment by rocallahan** [comment]
Also see if recording with `-n` makes a difference.

**Comment by zentner-kyle** [comment]
Here's the output of the debug session:
```
--------------------------------------------------
 ---> Reached target process 0 at event 170.
--------------------------------------------------
Remote debugging using 127.0.0.1:19874
warning: limiting remote suggested packet size (206696 bytes) to 16384
warning: Could not load shared library symbols for linux-vdso.so.1.
Do you need "set solib-search-path" or "set sysroot"?
Reading symbols from /home/zentner/rr/obj/bin/../lib/librrpreload.so...done.
Reading symbols from /usr/lib/libc.so.6...(no debugging symbols found)...done.
Reading symbols from /usr/lib/libdl.so.2...(no debugging symbols found)...done.
Reading symbols from /usr/lib/libpthread.so.0...(no debugging symbols found)...done.
Reading symbols from /lib64/ld-linux-x86-64.so.2...(no debugging symbols found)...done.
0x0000000070000038 in ?? ()
(gdb) p __pthread_force_elision
$1 = 1
(gdb) p __elision_available
$2 = 1
(gdb) p __elision_aconf
$3 = 3
(gdb) p (unsigned char[16]) __elision_aconf
$4 = "\003\000\000\000\003\000\000\000\003\000\000\000\003\000\000"
```

Recording with `-n` does make the trivial program replay successfully.

However, Firefox still does not replay successfully, failing with the same assertion (with `for 'SYSCALL: gettimeofday'`) at time 1607. The variables above have the same value when replaying this trace.

**Comment by rocallahan** [comment]
OK, so my guess was correct --- we failed to disable elision correctly.

Can you debug rr a bit for me? `Monkeypatcher::patch_after_mmap` in `monkeypatcher.cc` should be patching the third int in `__elision_aconf` to zero. Either that code is not getting hit at all, or we're failing to read the symbols, or somehow the symbols aren't present where we're looking for them, or we patch them but then they get reset for some reason.

**Comment by rocallahan** [comment]
I think I've fixed this with ad21f6ffd65b65823b8547ee029897a5f7dcb119.

**Comment by zentner-kyle** [comment]
I can't reproduce this on the latest master, and the cause is consistent. So yes, the above is correct.

Thanks!
